webui/react/src/services/stream/keyCache.ts

```diff
@@ -0,0 +1,82 @@
+export function decode_keys(keys: string): Array<number>{
+    const retval: Array<number> = [];
+    if (keys.length === 0) {
+        return retval;
+    }
+    for (const key of keys.split(',')) {
+        if (key.includes('-')) {
+            const [start, end] = key.split('-');
+            for (let i = Number(start); i <= Number(end); i++) {
+                retval.push(i);
+            }
+        } else {
+            retval.push(Number(key));
+        }
+    }
+    return retval;
+}
+
+export function encode_keys(set: Set<number>): string {
+    if (!set || set.size === 0) return '';
+    const out: Array<string> = [];
+    const keys = Array.from(set).sort((l, r) => l - r).values();
+    let start: number = keys.next().value;
+    let end = start;
+
+    const emit = (start: number, end: number, out: Array<string>) => {
+        if (start === end) {
+            out.push(start.toString());
+        } else {
+            out.push(`${start}-${end}`);
+        }
+    };
+
+    for (const k of keys) {
+        if (k === end + 1) {
+            end = k;
+            continue;
+        }
+        // end of a range
+        emit(start, end, out);
+        start = k;
+        end = start;
+    }
+    emit(start, end, out);
+
+    return out.join(',');
+}
+
+export class KeyCache {
+
+    #keys: Set<number>;
+    #maxseq: number;
+
+    constructor(keys?: Set<number>) {
+        this.#keys = keys || new Set();
+        this.#maxseq = 0;
+    }
+
+    #delete_one(id: number) {
+        this.#keys.delete(id);
+    }
+
+    public upsert(ids: Array<number>, seq?: number): void {
```

Comment:
why would there be an upsert without a `seq`?

Reply:
The current implementation supports traditional fetch for initial loading, then using stream to monitor changes, so when subscribing, there is a list of known (from traditional fetch) without `seq` number, but we could also make `seq` required and default to 0?

Reply:
I've removed the hybrid strategy. Regarding exposing the sync message to the consumer, I wonder what would be the use case of that?

Reply:
Imagine you are showing the trials tab in the experiment detail page.

You subscribe to trials matching the experiment id, and you render a skeleton while the data is loading.  When do you switch from the skeleton to the actual trials?

Answer: after you see a sync done message.  Otherwise, during the offline messages, you'd be really rapidly refreshing the page, first with one trial, then two, then three, etc, as the offline messages arrive.

Logically, this is analogous to the delay in the initial fetch with a normal polling strategy.  You can't load until you see the first fetch response.